StatusManager: consolidate status updates from different zones

npinaeva opened the pull request:
Create StatusManager - a centralized component responsible for updating
the centralized status of an object, based on zone-specific statuses.
Created as a part of cluster manager, handles only apbroutepolicy
objects for now.

Update AdminPolicyBasedRouteStatus.Messages to allow patching with
merge strategy. Update `update-codegen` script to always install
latest controller-gen, so that `controller-gen.kubebuilder.io/version`
on the generate object doesn't decrease.

Remove updated policy check based on timestamp,
since LastTransitionTime precision is in seconds, and the whole test
takes less than a second to complete, therefore all timestamps will
be the same for multiple updates. Just checking expected policy state
is enough for that test.

Update unit tests to check Status.Messages instead of Status.Status

Add distributed status management for EgressFirewall.
Add Status.Messages field to record statuses from zones,
make egressfirewall status a subresource

Be careful: requires CRD and status subresource permission change

npinaeva commented:
caught https://github.com/ovn-org/ovn-kubernetes/issues/3924

trozet commented on go-controller/hack/update-codegen.sh:
```diff
@@ -9,16 +9,13 @@ if [ -z "${crds}" ]; then
+olddir="${PWD}"
+builddir="$(mktemp -d)"
+cd "${builddir}"
+GO111MODULE=on go install sigs.k8s.io/controller-tools/cmd/controller-gen@latest
```
thanks

trozet commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,122 @@
+	_, err := m.patch(context.TODO(), newObj.GetNamespace(), newObj.GetName(), ktypes.ApplyPatchType, patch, patchOptions, "status")
+
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("could not patch: %w", err))
```
do you want to list the object namespaced name that cannot be patched? or are you expected that to be included in the err?

trozet commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+type StatusManager struct {
+	abpRouteManager *typedStatusManager[*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute]
```
typo: apb

maybe it should be apbStatusManager?

trozet commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+func (sm *StatusManager) Start() error {
+	klog.Infof("Starting StatusManager with params: abpRouteManager=%v",
```
maybe we could make this function more generic by using reflect to get the fields of the struct and iterate through each one?

npinaeva replied:
yeah that's easy

jcaamano replied:
I don't think it helped enough :/

This code is bound to grow with more statuses being managed and we will keep adding and in-lining more code.

If you want to stick to generics, we could have an interface that could be implemented on a separate file for each type
```
type statusHandler[T any] interface {
    get ...
    patch ...
    ...
}
```

And then another type with the common logic
```
type statusController[T statusHandler[T]] struct {
    statusHandler statusHandler[T]
    informer ...
    ...
}
```

Just an idea

npinaeva replied:
I tried something like that, but now I killed lint, because it can't find interface usages. But the code should work

npinaeva replied:
I think the lint failure is this https://github.com/dominikh/go-tools/issues/1440
I may need to switch back to inlining instead of using an interface, until the bug is fixed, @jcaamano wdyt?

trozet commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,122 @@
+	metav1.Object
+}
+
+func (m *typedStatusManager[T]) updateStatus(oldObjInterface, newObjInterface interface{}) {
```
so updateStatus returns no error, dont we want to return error and add the key back to a workqueue and retry updating status?

npinaeva replied:
yeah that will complicate the code a bit
I thought that failing to update the status doesn't necessarily need a retry, but if you think it's worth it, I can add a queue

trozet commented on go-controller/pkg/ovn/controller/apbroute/external_controller_policy_test.go:
```diff
@@ -821,3 +804,23 @@ var _ = Describe("OVN External Gateway policy", func() {
 	})
 })
+
+func eventuallyCheckAPBRouteStatus(policyName string, expectFailure bool) {
```
nice

trozet commented on go-controller/pkg/ovn/controller/apbroute/master_controller.go:
```diff
@@ -100,32 +104,49 @@ func (c *ExternalGatewayMasterController) GetAdminPolicyBasedExternalRouteIPsFor
+		Force:        utilpointer.Bool(true),
+		FieldManager: c.zoneID,
+	}
+	patch := []byte(fmt.Sprintf(`{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute","status":{"lastTransitionTime":"%s","messages":["%s"]}}`, metav1.Now().ToUnstructured().(string), newMsg))
```
so I see you doing server side apply and then identifying each manager by zone id, but I dont get if each node is updating the same field? Like are they all updating the same status field? Maybe some example output will help me understand.

jcaamano replied:
It is updating a field that is an array defined as a set with a merge patch strategy. So basically each node adds an entry to the list instead of replacing the whole list. I think the purpose is to have a single entry per node but I don't know what is the key of the set. If the key is the whole message, a single node could add multiple entries to the list if the error changes or if we go from error to success, instead of replacing its previous entry.

jcaamano replied:
I thought the plan here was to use conditions where the condition type would be unique for each node. Similar to how ANP is doing it I think.

npinaeva replied:
this whole patch strategy is a bit tricky, so here is what you will find in the managedFields
```
    managedFields:
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:status: {}
      manager: cluster-manager
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-control-plane configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-control-plane
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-worker configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-worker
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
    - apiVersion: k8s.ovn.org/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:lastTransitionTime: {}
          f:messages:
            'v:"ovn-worker2 configured external gateway IPs: 1.1.1.1"': {}
      manager: ovn-worker2
      operation: Apply
      subresource: status
      time: "2023-10-25T11:04:16Z"
```
every field has its manager, which is set by `patchOptions.FieldManager` for patch call, so every node owns its message in the list.

`Server-Side Apply tries to merge fields based on the actor who manages them instead of overruling based on values. This way multiple actors can update the same object without causing unexpected interference.`  https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy, so what happens is every owner has its elements in the messages list (potentially multiple elements, but in this case only 1). Now every patch from that owner is treated as a new state for all the list elements owned by this manager. This way every time we patch messages with only 1 message, all messages owned by a given manager will be replaced with this 1 message.
There are some nice examples in the tests for merge code, e.g. https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/set_test.go#L502-L541.

There is still a problem of cleaning up messages owned by deleted nodes/zones. We could probably make some cleanup on cluster-manager start to check which zones are present, and delete all non-existing zone messages. I can try to add that part once we agree on the general framework here.

jordigilh replied:
Looks great! I wonder what will the client display when the user runs the `kubectl get apbexternalroute` command, since each zone will have its own status recorded.

jcaamano replied:
Yeah, I wonder the same thing, how it would it look and perform to have 250 nodes (or more) include their message into each resource. Another approach would be "no news, good news" and just add a message for nodes that have had a failure which of course is not the same thing so it would be a compromise.

npinaeva replied:
just `get` will only show `status.status` (cumulative), and all messages are only shown with `-o yaml`

jcaamano commented:
Missing a bit of documentation somewhere on the general approach that guides future implementations for other resource statuses. It might be confusing that, while we could be doing different things for apb and ef because the code structure supports it, we are doing basically the same thing for both.

npinaeva commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,156 @@
+			break
+		}
+	}
+	return []byte(fmt.Sprintf(`{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute","status":{"status":"%s"}}`, newStatus)),
```
From what I know, not really. I tried marshalling objects, but in that case all fields of Message object will be added to the patch, but we don't really want that, since cluster manager only changes Status.Status and nodes only change Status.Messages and Status.LastTransitionTime. The problem is that nil/empty fields are treated by patch as a request to delete/reset the field, and even though you can't reset fields owned by other managers, the request doesn't really express what I want to happen. Another problem is that it breaks unit tests, since test apiserver is not smart enough to evaluate ownership conflicts, and cleans up unset fields :) 
Another way to do that is to create a special object to represent the status with desired json marshalling, but just using a string looks much simpler to me. lmk what you think

jcaamano replied:
ummm, isn't that a problem of our CRDs? We are not using pointers and `omitempty` as we should? Maybe we can change that now that we are at it if it is backward compatible.

npinaeva replied:
not really. because patching is changing a part of CRD, e.g. you may want to omit some required fields, just because you don't want to change them with a patch, while CRD will define that field as required and shouldn't be omitted.

jcaamano replied:
OK. But when we patch other things elsewhere I don't see us including the `{"apiVersion":"k8s.ovn.org/v1","kind":"AdminPolicyBasedExternalRoute"` part. Why would we need that here?

npinaeva replied:
ok, it wasn't easy, but I regenerated all clients :D
check last 2 commits to see how it looks, I left them separate to potentially backport only the first 2

npinaeva commented on go-controller/pkg/types/const.go:
```diff
@@ -197,4 +197,7 @@ const (
 	// may be refreshed with a new ARP request.
 	GRMACBindingAgeThreshold = "300"
+
+	APBRouteErrorMsg         = "failed to apply policy"
+	EgressFirewallApplyError = "EgressFirewall Rules not correctly applied"
```
I am not sure if defining those as `error` is useful, because those are never used in the code as `error`. Usually there is another `error` returned by the handler, and these messages are just inserted to the status message for more context. I could make some `error_msg` file and put all error messages there maybe?

jcaamano replied:
maybe status.go?

jcaamano replied:
or resource_status.go

npinaeva commented:
For better visibility: I think the lint failure is this https://github.com/dominikh/go-tools/issues/1440
I will probably add one more commit to make linter happy in the end hoping to revert it when the bug is fixed

jordigilh commented on go-controller/pkg/clustermanager/status_manager/resource_manager.go:
```diff
@@ -0,0 +1,68 @@
+		return nil, false
+	}
+	newStatus := adminpolicybasedrouteapi.SuccessStatus
+	for _, message := range route.Status.Messages {
```
Is this iterating over the latest message from each zone or does it go through all previous messages as well? Maybe I'm confused and what really happens is that each zone is only updating its single entry in the set.

npinaeva replied:
that is exactly what happens, every zone owns one message in the list with the latest status

jcaamano commented on go-controller/pkg/ovn/egressfirewall.go:
```diff
@@ -294,6 +295,7 @@ func (oc *DefaultNetworkController) addEgressFirewall(egressFirewall *egressfire
 	defer ef.Unlock()
 	// egressFirewall may already exist, if previous add failed, cleanup
 	if _, loaded := oc.egressFirewalls.Load(egressFirewall.Namespace); loaded {
+		klog.Infof("Egress firewall in namespace 5s already exists, cleanup", egressFirewall.Namespace)
```
typo: `5s`

jcaamano commented on go-controller/pkg/ovn/controller/apbroute/master_controller.go:
```diff
@@ -100,32 +103,51 @@ func (c *ExternalGatewayMasterController) GetAdminPolicyBasedExternalRouteIPsFor
+	}
+	newMsg := fmt.Sprintf("configured external gateway IPs: %s", strings.Join(sets.List(gwIPs), ","))
+	if syncError != nil {
+		newMsg = fmt.Sprintf("%s %s: %v", c.zoneID, types.APBRouteErrorMsg, syncError.Error())
```
Aren't we printing the zone ID twice in the message?

jcaamano commented on go-controller/go.mod:
```diff
@@ -58,6 +58,7 @@ require (
 	kubevirt.io/api v1.0.0-alpha.0
 	sigs.k8s.io/controller-runtime v0.15.1
 	sigs.k8s.io/network-policy-api v0.1.1
+	sigs.k8s.io/structured-merge-diff/v4 v4.2.3
```
Out of curiosity, what does this do?

npinaeva replied:
it is used by `applyconfiguration`, e.g. here https://github.com/ovn-org/ovn-kubernetes/pull/3750/commits/5c707cdd6d6d6d992c5f93e8f1f9f498622ae702#diff-a56cb620ea0baa29f6de1f61806fecded83591d3c2339ebac00e7da47411ca54R24

jcaamano replied:
I can't see this anymore after the latest push

npinaeva replied:
that's because it points to the collapsed file :( check [go-controller/pkg/crd/adminpolicybasedroute/v1/apis/applyconfiguration/internal/internal.go](https://github.com/ovn-org/ovn-kubernetes/commit/5c707cdd6d6d6d992c5f93e8f1f9f498622ae702#diff-a56cb620ea0baa29f6de1f61806fecded83591d3c2339ebac00e7da47411ca54)

jcaamano commented on test/e2e/status_manager.go:
```diff
@@ -0,0 +1,120 @@
+		}
+	})
+
+	ginkgo.It("Should validate the egress firewall status when adding an unknown zone", func() {
```
is this a way to test status manager through egress firewall status, or are we missing tests for APB route?

npinaeva replied:
yeah this one is intended for status manager testing, doesn't really matter which object to use.
I will try to add all resource-specific tests to unit tests.

npinaeva replied:
now I think I probably will add 1 extra test for every resource to catch potential rbac problems downstream

npinaeva replied:
morning thoughts: we don't need to add resource-specific tests, because both egress firewall and apbroute e2e tests check their statuses to figure out when the config is applied (and hopefully the same will apply to the other resources in the future, at least it seems to be the only way to avoid races)

npinaeva commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,277 @@
+
+	applyOptions := &metav1.ApplyOptions{
+		Force:        true,
+		FieldManager: fieldManagerName,
```
I intended to show the meaning of this name (which is field manager) and considered it is a part of cluster manager package, adding cluster manager would be excessive.
Do you think `clusterManagerFieldManagerName` will be better? Or I can also rename it to `clusterManagerName` if you find it descriptive enough

jcaamano replied:
I think `clusterManagerName` works good enough, this being a non-exported constant.

jcaamano commented on go-controller/pkg/clustermanager/clustermanager.go:
```diff
@@ -145,6 +146,11 @@ func (cm *ClusterManager) Start(ctx context.Context) error {
 		}
 	}
+
+	if err := cm.statusManager.Start(cm.zoneClusterController.zoneTracker.Subscribe); err != nil {
```
I guess that what I want to say is that is weird that we involve the `zoneClusterController`, which has nothing to do with what you are zone here, just because of the convenience that is already handling node updates or it has zone in the name.

It is also using the retry framework instead of being level driven design, so that is adding more to that pile.

I think what would make sense is to do a zone tracker that tracks node updates by itself. It can be private to the status manager package. If we need to reuse it someday, we could promote it to its own package and maybe give it a subcription interface.

npinaeva replied:
Ah I see what you mean.
I agree that zoneTracker should have its own handler, which I will add. But I still would like to have it as a separate package inside status_manager, in case we need to move it somewhere later, and also it is a logically separate controller that I believe makes sense to have in an isolated package

jcaamano commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,135 @@
+	controller := level_driven_controller.NewStoppableController[T](name, controllerConfig)
+	return &typedStatusManager[T]{
+		name: name,
+
```
nit: white line

jcaamano commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -0,0 +1,135 @@
+type typedStatusManager[T any] struct {
+	name string
+
+	objController level_driven_controller.StoppableController[T]
```
We are just doing start & stop so objController does not need to be typed

npinaeva replied:
yeah I could define an interface for both controller types, but it is not really "required" here, and I also prefer to not use interfaces unless there are 2 different implementations for it, since it complicates the code in my opinion.
e.g. I find https://github.com/ovn-org/ovn-kubernetes/pull/3750/files#diff-25788e003f7180dde6dbf91380e451e092a7f0b67e850209ac884b15fc4973ffR54 easier to read and navigate with the typed generic compared to an interface, not a strong opinion though :)

jcaamano replied:
I just think there should only be one.

And iI am of a different opinion :P, generics code where types are not inferred is much harder to read than interfaces.

npinaeva replied:
the type is inferred here, its same as `typedStatusManager` type.
Interface is an extra layer of abstraction, that I would say should be avoided if you can use specific type. Generic controller is like a regular controller, it doesn't require an interface by itself unless you need some controllers manager that should work with any controller type.

jcaamano replied:
I mean in general. Reading `...[...]` its never better than just `...`, whatever it is, IMHO.

Interfaces are THE abstraction that should never be avoided, also IMHO.

Anyhow, I am not pushing for the interface really. It is not that important here since the controller is just an utility. It is just that I saw this first, before seeing that you had to ways to start and stop, so I when I saw this I thought your intention was something more akin to an interface. You can ignore this one.

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,268 @@
+package level_driven_controller
```
yeah I know about that nice lib, but I faced 2 main problems with it:
1. it uses manager to create multiple controller, and a manager starts its own watchers. But for perf reasons, we want to reuse our watch factory, so as I understand it, we need to switch all our controllers to use the same manager to have the minimal amount of watchers.
2. there is no initialSync functionality (again, maybe there is a workaround for that case, but I didn't figure it out), which we use pretty often on startup.
My main goal was not to have a "good level-driven controller", but mostly to generalize our existing code for level-driven controllers (since I had to use it twice in this PR) that we use in ovn-k today. If I get to choose, I would probably prefer switching to controller-runtime instead of improving this controller to be of comparable quality.

I don't think it can be inside util, because this package imports `factory`, and factory import `util`.

jcaamano replied:
`util/controller` would be it's own package, not imported by factory.

but probably we could think of a new package as `<newpkg>/factory`, `<newpkg/informer>`, `<newpkg>/controller`

npinaeva replied:
new package makes sense, but probably will be good for another PR
I will just rename new pkg to controller for now

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,271 @@
+	// In case some events shouldn't be handled, we can skip them with this config.
+	// Add and Update are configured by one variable, because on restart all existing objects will trigger add event,
+	// which in reality may be an update copared to the previously handled events.
+	HandleAddUpdate bool
```
Yep, check this https://github.com/ovn-org/ovn-kubernetes/pull/3750/files#diff-15125dcc990bd208c39c356846cbea3edb21b8bfee37b166ca38a273e4ccfed1R101 (if the link doesn't work again, search for WithUpdateHandlingForObjReplace)

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,271 @@
+	// Add and Update are configured by one variable, because on restart all existing objects will trigger add event,
+	// which in reality may be an update copared to the previously handled events.
+	HandleAddUpdate bool
+	HandleDelete    bool
```
it can, I just tried to save some hustle to queue and reconcile something that we know doesn't need to be reconciled (as you can just omit onDelete function with `AddHandler`)
I can remove it, or I can change it to `skipAdd/Update/Delete` so that default value is false, and you don't need to mention it unless you are trying to skip some event, wdyt?

jcaamano replied:
I would just remove it. API wise it complicates it seems unnecessarily if a user is still empowered to make those decisions in Reconcile.

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,271 @@
+	return true
+}
+
+func (c *controller[T]) AddToQueue(key string) {
```
It is used by status manager when zones change to reconcile all objects. I think generally speaking something like this makes sense, but I am not sure what is better `AddToQueue` or `ReconcileAll`

jcaamano replied:
OK. 

So this is a way to kind of connect or compose a sub-controllers into a single controller. Just as the library I pointed before, same capability but a slightly different approach, where you have a single controller that can watch multiple things.

Might look a bit weird to people accustomed to see traditional controllers, and who knows what kind of complications this might lead us, but nothing we haven't done before so I guess this is how we will roll for now.

It looks like ReconcileAll could be done here and It would save of us from some code repetition?

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,271 @@
+
+// StoppableController is a level-driven controller that is shut down after Stop() call.
+// It creates its own stop channel and waiting group to ensure proper cleanup.
+type StoppableController[T any] struct {
```
in ovn-k we have 2 ways to stop a controller: with stop channel or with Stop() call. These options require different Start() signatures, so to allow multiple `Start` options, I made these 2 types (there could be a third one using context in the future). To reuse start functionality, I had to create a common `start` method, and then 2 wrapper types to call it with different args.

jcaamano replied:
I think we should have a single way for this API and users should adapt to it. So instead of encouraging all the different ways you can handle this, let's just go for one.

npinaeva replied:
let me try to clarify what i meant with this package
It is not supposed to be a better way to write controllers, but just a way to not copy the same code multiple times, mostly to avoid bugs. That is why in this package I am trying to adapt to how controllers are used right now (that includes implementing Stop() methods making sure the channel is closed and Wait() call is there) instead of defining the nice interface and then catching bugs from the user side.
I find it useful even if it is not perfect, since it decreases code duplication, and helps you use the controller in a proper way (e.g. you can find some strange combinations of wait group, stop channel and Stop method usages, so this interface is supposed to help with that)

jcaamano replied:
The only controllers using this utility are the ones introduced in this PR and they just use `NewStoppableController`.
 
Since it is only that, you wouldn't need a new type (unless you want an interface). I also think we should start to avoid if possible stop channels and moving wait groups around. So I would just have a `Run(ctx Context, ...)` that blocks. Why would we encourage different ways of doing it? We should consolidate upstream users to do it in the same way, and If it can't be avoided, they can have their own adapters. And if it can't be avoided, those adapters can also be utilities in their own package with their own doc. 

`NewStopChanController` is just used in the test so why not move it to the test?

It is very confusing to export `NewStopChanController` and `NewStoppableController`

I am just trying to help, within reason, but if you are not convinced then I guess I failed and we can just move on with what you have.

npinaeva commented on go-controller/pkg/level_driven_controller/level_driven_controller.go:
```diff
@@ -0,0 +1,271 @@
+	wg            *sync.WaitGroup
+}
+
+func NewStoppableController[T any](name string, config *ControllerConfig[T]) StoppableController[T] {
```
since I have 2 types: `StoppableController` and `StopChanController` I am not sure what is the best naming here.
We could say e.g. that `StopChanController` is the default one, and call it Controller, with New method

jcaamano replied:
The fact that you have two is in itself confusing. We should just have one and users should deal with it.

jcaamano commented on go-controller/pkg/types/resource_status.go:
```diff
@@ -1,7 +1,20 @@
+
 // this file defines error messages that are used to figure out if a resource reconciliation failed
 const (
 	APBRouteErrorMsg         = "failed to apply policy"
```
It would make sense for this to follow the same naming: `ErrMsg` vs `ApplyError`

jcaamano commented on go-controller/pkg/clustermanager/status_manager/zone_tracker/zone_tracker.go:
```diff
@@ -0,0 +1,239 @@
+	return true
+}
+
+// Subscribe allows to get notified about zones changes (not nodes changes,
+// e.g. if a new node for existing zone is added, callbacks won't be executed).
+// First callback will be executed when zoneTracker finishes its initial sync, that may happen immediately.
+// Only 1 subscriber is allowed for now, therefore an error will be returned on the second call.
+func (zt *ZoneTracker) Subscribe(onZonesUpdate func(newZones sets.Set[string])) error {
+	zt.zonesLock.Lock()
+	defer zt.zonesLock.Unlock()
+	if zt.onZonesUpdate != nil {
+		return fmt.Errorf("ZoneTracker allows only 1 subscriber")
+	}
+	zt.onZonesUpdate = onZonesUpdate
+	zt.notifySubscriber()
+	return nil
+}
+
+// notifySubscriber must be called with zonesLock
+func (zt *ZoneTracker) notifySubscriber() {
+	if zt.inited && zt.onZonesUpdate != nil {
+		zt.onZonesUpdate(sets.New[string](zt.zones.UnsortedList()...))
+	}
+}
```
This is just a nit, but if there can ever be just one subscriber, why just not pass the user callback in the zone tracker constructor. Otherwise the double indirection is just confusing and seems more complicated that needs be

jcaamano commented on go-controller/pkg/clustermanager/status_manager/zone_tracker/zone_tracker.go:
```diff
@@ -0,0 +1,239 @@
+	}
+
+	zoneDeleted := zt.deleteNodeZoneNoLock(nodeName)
+	// notify subscriber if the zone was deleted or added (zone is added if it wasn't in the zones set, and we are adding it in 2 lines)
```
`adding it in 2 lines`? can't understand :P

npinaeva commented on go-controller/pkg/clustermanager/status_manager/zone_tracker/zone_tracker.go:
```diff
@@ -0,0 +1,239 @@
+	if zoneName == UnknownZone {
+		// wait for unknownZoneTimeout to allow node to get a zone label.
+		// if that doesn't happen, consider a node is not a part of any zone and remove
+		go func() {
+			select {
+			case <-zt.stopChan:
+				return
+			case <-time.After(zt.unknownZoneTimeout):
+				// if node still doesn't have a zone, delete
+				zt.zonesLock.Lock()
+				defer zt.zonesLock.Unlock()
+
+				if zt.nodeToZone[nodeName] == UnknownZone {
+					klog.Warningf("Node %s has UnknownZone for more than %v", nodeName, zt.unknownZoneTimeout)
+					if zt.deleteNodeZoneNoLock(nodeName) {
+						zt.notifySubscriber()
+					}
+					zt.unknownZoneNodes.Insert(nodeName)
+				}
+			}
+		}()
```
what is the problem with blocking for a timeout? Also not sure how it can based on a timestamp

jcaamano replied:
Oh ignore, it's a gorutine.

jcaamano replied:
I would still consider either of my other two suggestions though

jcaamano replied:
There might be some other weird scenarios, like a node being added and then removed while the goroutine for the first event is still running.

npinaeva commented on go-controller/pkg/clustermanager/status_manager/status_manager.go:
```diff
@@ -66,50 +89,105 @@ func (m *typedStatusManager[T]) Stop() {
+		}
+
+		// now calculate accumulated status.
+		// if not all zones reported status, clean it up, since the status is considered unknown until all zone report results.
```
yeah we can set failed status without having all zone messages, I just didn't think it is important enough to reconcile for that case, but I can add it

npinaeva replied:
I also didnt want to make resourceManagers zone-aware, so I will need to pass an extra flag to only apply status if it is failure

jcaamano replied:
Yeah, since it is private to the package it is not a big deal. I guess it wouldn't be a big deal either passing the zones that the resourceManagers should expect a message from, so they filter out messages that should not be taken into account.

npinaeva commented:
Changes gist:
- renamed EgressFirewallApplyError to EgressFirewallErrorMsg
- renamed level_driven_controller to controller
- left only 1 Start() method for Controller and an interface
- unified Config and HandlerFuncs in controller pkg to be one generic type that is passed to constructor
- added timestamp-based tracking for zone tracking unknown zones

jcaamano commented on go-controller/pkg/controller/controller.go:
```diff
@@ -0,0 +1,217 @@
+type Controller interface {
+	Start(threadiness int) error
+	Stop()
+	ReconcileAll()
```
nit: The `ReconcileAll` looks a bit out of place now. But I guess this should evolve to have a `Reconcile` as  well as some other methods enable watching different types of resources so not bad from that perspective.

npinaeva replied:
I expected ReconcileAll() to be moved to the resource handler interface, when one controller will allow multiple resources, so that you can specify which resource should be reconciled, if that makes sense

npinaeva commented on go-controller/pkg/controller/controller.go:
```diff
@@ -0,0 +1,217 @@
+	ReconcileAll()
+}
+
+type Config[T any] struct {
```
right, I should add a comment for that. Considering this controller can be extended to handle multiple resources (each resource will be creates with Config), Initial sync should be called only once (not per-resource) before workers are started (e.g. https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/controller/egressservice/egressservice_zone.go#L229)
So I kept it separate to make less changes in the future, but I can also make it a part of the Config for now if you'd like that more?

jcaamano commented on go-controller/pkg/clustermanager/status_manager/zone_tracker/zone_tracker.go:
```diff
@@ -0,0 +1,232 @@
+		zt.unknownZoneNodes[nodeName] = time.Now()
+		// wait for unknownZoneTimeout to allow node to get a zone label.
+		// if that doesn't happen, consider a node is not a part of any zone and remove
+		go func() {
+			select {
+			case <-zt.stopChan:
+				return
+			case <-time.After(zt.unknownZoneTimeout):
+				zt.checkUnknownNodeTimeout(nodeName)
+			}
+		}()
```
I read around that timers might be not that accurate depending on the environment.
https://stackoverflow.com/questions/51415965/about-the-accuracy-of-the-time-timer
I haven't had time to find a formal reference.

The timeout for that persistent thread to start evaluating `zt.unknownZoneNodes` again could be based on the diff between the current time and the earliest expiration found in `zt.unknownZoneNodes` on the last evaluation.

You can add that extra second as well.

npinaeva replied:
I think timer inaccuracy problem will be present in both cases, so I just added a little delta to the condition check, I think that should be sufficient

jcaamano replied:
Yeah, it is something to consider in both alternatives.

I guess that what my alternative has going for it is single thread vs multiple threads as well as a more reassuring way to keep track that we eventually process all that we need to process from `zt.unknownZoneNodes`.

But I will take your approach as well.
